File: edp_java/src/main/Waiter.java
package main;

import java.util.Optional;
import java.util.concurrent.Semaphore;

public class Waiter extends Thread {
    // Constants (Time in milliseconds)
    private static final int CHECK_ORDERS_INTERVAL = 500;

    // Static
    private static final Waiter[] waiters = java.util.Arrays.stream(Name.WAITER_NAMES)
            .map(n -> new Waiter(n.toString())).toArray(Waiter[]::new);

    private static final int NUM_COOKS = Cook.list().length;
    public static final Semaphore cookSemaphore = new Semaphore(NUM_COOKS, true);

    // Synchronized queue for philosophers calling waiters
    private static final java.util.concurrent.BlockingQueue<Philosopher> WAITER_CALLING_QUEUE = new java.util.concurrent.LinkedBlockingQueue<>();

    public static Waiter[] list() {
        return waiters;
    }

    public static final Waiter availableWaiter() {
        for (Waiter waiter : waiters) {
            if (!waiter.isProcessingOrder()) {
                return waiter;
            }
        }
        return null; // No available waiter found
    }

    public static void addPhilosopherToQueue(Philosopher philosopher) {
        try {
            WAITER_CALLING_QUEUE.put(philosopher);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    public static void removePhilosopherFromQueue(Philosopher philosopher) {
        WAITER_CALLING_QUEUE.remove(philosopher);
    }

    // Instance
    private Name name;
    private Optional<Order> order = Optional.empty();
    private Optional<Philosopher> patron = Optional.empty();
    private boolean isProcessingOrder = false;

    public Waiter(String name) {
        this.name = new Name(name);
    }

    public String name() {
        return name.toString();
    }

    private boolean isProcessingOrder() {
        return isProcessingOrder;
    }

    public void takeOrder(Philosopher philosopher, Order order) {
        this.isProcessingOrder = true;
        this.order = Optional.of(order);
        this.patron = Optional.of(philosopher);
    }

    @Override
    public void run() {
        System.out.println(String.format("Waiter %s is ready to take orders.", name));
        while (!Thread.currentThread().isInterrupted()) {
            try {
                if (isProcessingOrder && order.isPresent() && patron.isPresent()) {
                    processOrder();
                } else {
                    // Check for completed meals to deliver
                    if (Counter.hasCompletedMeals()) {
                        Order completedOrder = Counter.pollCompletedMeal();
                        if (completedOrder != null) {
                            deliverOrder(completedOrder);
                        }
                    } else {
                        // No completed meals available, check for philosophers who need service
                        checkForPhilosophersNeedingService();
                    }
                }

                Thread.sleep(CHECK_ORDERS_INTERVAL);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }
    }

    private void processOrder() {
        if (!order.isPresent() || !patron.isPresent()) {
            return;
        }
        Order currentOrder = order.get();
        Philosopher currentPatron = patron.get();
        boolean cookAcquired = cookSemaphore.tryAcquire();
        if (cookAcquired) {
            boolean orderPlaced = false;
            try {
                Cook.ORDER_QUEUE.put(currentOrder);
                orderPlaced = true;
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            if (orderPlaced) {
                System.out.println(
                        String.format("Waiter %s placed order for %s.", name, currentPatron.name()));
            }
        } else {
            System.out
                    .println(String.format("Waiter %s cannot place order for %s - all chefs busy. Giving $5.00 coupon.",
                            name, currentPatron.name()));
            currentPatron.setShouldReceiveCoupon(true);
            currentPatron.notifyFoodServed(); // Countdown the latch to wake up the philosopher
            // Clear the philosopher's order so they can try again
            currentPatron.clearOrder();
            // Vacate the seat and print message
            currentPatron.vacateSeat();
            System.out.println(String.format("Philosopher %s has left the restaurant without being served.",
                    currentPatron.name()));
        }
        order = Optional.empty();
        patron = Optional.empty();
        isProcessingOrder = false;
    }

    private void deliverOrder(Order completedOrder) {
        Philosopher targetPhilosopher = completedOrder.philosopher();
        targetPhilosopher.notifyFoodServed();
        System.out.println(String.format("Waiter %s is serving philosopher %s %s.", name,
                targetPhilosopher.name(), completedOrder.mealString()));
        // Make waiter available for new orders
        order = Optional.empty();
        patron = Optional.empty();
        isProcessingOrder = false;
    }

    private void checkForPhilosophersNeedingService() {
        if (isProcessingOrder) {
            return; // Already processing an order
        }

        // Check the synchronized queue for philosophers calling waiters
        Philosopher callingPhilosopher = WAITER_CALLING_QUEUE.poll();

        if (callingPhilosopher != null) {
            // Philosopher needs service
            Order newOrder = new Order(callingPhilosopher);
            callingPhilosopher.setOrder(newOrder);
            takeOrder(callingPhilosopher, newOrder);

            String mealDesc = newOrder.mealString();
            System.out.println(String.format("Waiter %s has taken order for %s from philosopher %s.",
                    name, mealDesc, callingPhilosopher.name()));
        }
    }

    // class implementation
    public void serve(Philosopher philosopher) {
        if (order.isPresent()) {
            System.out.println(String.format("Waiter %s is serving philosopher %s %s.", name,
                    philosopher.name(), order.get().mealString()));
        }
    }
}

File: edp_go/waiter.go
package main

import (
	"context"
	"fmt"
	"sync"
	"time"
)

type WaiterManager struct {
	waiters          []*Waiter
	philosopherQueue chan *Philosopher
	cookSemaphore    chan struct{}
}

var waiterManager = &WaiterManager{
	philosopherQueue: make(chan *Philosopher, 1000),
	cookSemaphore:    make(chan struct{}, len(COOK_NAMES)),
}

func init() {
	for i := 0; i < len(COOK_NAMES); i++ {
		waiterManager.cookSemaphore <- struct{}{}
	}
	waiterManager.waiters = make([]*Waiter, len(WAITER_NAMES))
	for i, name := range WAITER_NAMES {
		waiterManager.waiters[i] = NewWaiter(name)
	}
}

func (wm *WaiterManager) List() []*Waiter {
	result := make([]*Waiter, len(wm.waiters))
	copy(result, wm.waiters)
	return result
}

func (wm *WaiterManager) AvailableWaiter() *Waiter {
	for _, waiter := range wm.waiters {
		if !waiter.IsProcessingOrder() {
			return waiter
		}
	}
	return nil
}

func (wm *WaiterManager) AddPhilosopherToQueue(philosopher *Philosopher) {
	select {
	case wm.philosopherQueue <- philosopher:
	default:
	}
}

type Waiter struct {
	name              Name
	order             *Order
	patron            *Philosopher
	isProcessingOrder bool
	mutex             sync.RWMutex
	ctx               context.Context
	cancel            context.CancelFunc
	wg                sync.WaitGroup
}

func NewWaiter(name Name) *Waiter {
	ctx, cancel := context.WithCancel(context.Background())
	return &Waiter{name: name, ctx: ctx, cancel: cancel}
}

func (w *Waiter) Name() string { return string(w.name) }
func (w *Waiter) IsProcessingOrder() bool {
	w.mutex.RLock()
	defer w.mutex.RUnlock()
	return w.isProcessingOrder
}

func (w *Waiter) TakeOrder(philosopher *Philosopher, order *Order) {
	w.mutex.Lock()
	w.isProcessingOrder = true
	w.order = order
	w.patron = philosopher
	w.mutex.Unlock()
}

func (w *Waiter) Run() {
	defer w.wg.Done()
	fmt.Printf("Waiter %s is ready to take orders.\n", w.name)
	for {
		select {
		case <-w.ctx.Done():
			return
		default:
			w.mutex.RLock()
			processing := w.isProcessingOrder
			order := w.order
			patron := w.patron
			w.mutex.RUnlock()

			if processing && order != nil && patron != nil {
				w.processOrder()
			} else {
				if counter.HasCompletedMeals() {
					completedOrder := counter.PollCompletedMeal()
					if completedOrder != nil {
						w.deliverOrder(completedOrder)
					}
				} else {
					w.checkForPhilosophersNeedingService()
				}
			}
			time.Sleep(CHECK_ORDERS_INTERVAL)
		}
	}
}

func (w *Waiter) processOrder() {
	w.mutex.RLock()
	currentOrder := w.order
	currentPatron := w.patron
	w.mutex.RUnlock()

	if currentOrder == nil || currentPatron == nil {
		return
	}

	select {
	case <-waiterManager.cookSemaphore:
		select {
		case cookManager.orderQueue <- currentOrder:
			fmt.Printf("Waiter %s placed order for %s.\n", w.name, currentPatron.Name())
		default:
		}
	default:
		fmt.Printf("Waiter %s cannot place order for %s - all chefs busy. Giving $5.00 coupon.\n",
			w.name, currentPatron.Name())
		currentPatron.SetShouldReceiveCoupon(true)
		currentPatron.NotifyFoodServed()
		currentPatron.ClearOrder()
		currentPatron.VacateSeat()
		fmt.Printf("Philosopher %s has left the restaurant without being served.\n", currentPatron.Name())
	}

	w.mutex.Lock()
	w.order = nil
	w.patron = nil
	w.isProcessingOrder = false
	w.mutex.Unlock()
}

func (w *Waiter) deliverOrder(completedOrder *Order) {
	targetPhilosopher := completedOrder.Philosopher()
	targetPhilosopher.NotifyFoodServed()
	fmt.Printf("Waiter %s is serving philosopher %s %s.\n", w.name,
		targetPhilosopher.Name(), completedOrder.MealString())

	w.mutex.Lock()
	w.order = nil
	w.patron = nil
	w.isProcessingOrder = false
	w.mutex.Unlock()
}

func (w *Waiter) checkForPhilosophersNeedingService() {
	w.mutex.RLock()
	processing := w.isProcessingOrder
	w.mutex.RUnlock()

	if processing {
		return
	}

	select {
	case callingPhilosopher := <-waiterManager.philosopherQueue:
		newOrder := NewOrder(callingPhilosopher)
		callingPhilosopher.SetOrder(newOrder)
		w.TakeOrder(callingPhilosopher, newOrder)

		mealDesc := newOrder.MealString()
		fmt.Printf("Waiter %s has taken order for %s from philosopher %s.\n",
			w.name, mealDesc, callingPhilosopher.Name())
	default:
	}
}
